rclcpp/include/rclcpp/experimental/intra_process_manager.hpp

```diff
@@ -384,33 +405,71 @@ class IntraProcessManager
 
-      auto subscription = std::static_pointer_cast<
-        rclcpp::experimental::SubscriptionIntraProcess<MessageT>
-        >(subscription_base);
-
-      if (std::next(it) == subscription_ids.end()) {
-        // If this is the last subscription, give up ownership
-        subscription->provide_intra_process_message(std::move(message));
+      // same message type, so it can just be passed through
+      if (subscription_base->is_serialized() == is_serialized_publisher) {
+        auto subscription = std::static_pointer_cast<
+          rclcpp::experimental::SubscriptionIntraProcess<MessageT>
+          >(subscription_base);
+
+        if (std::next(it) == subscription_ids.end()) {
+          // If this is the last subscription, give up ownership
+          subscription->provide_intra_process_message(std::move(message));
+        } else {
+          // Copy the message since we have additional subscriptions to serve
+          MessageUniquePtr copy_message;
+          Deleter deleter = message.get_deleter();
+          auto ptr = MessageAllocTraits::allocate(*allocator.get(), 1);
+          MessageAllocTraits::construct(*allocator.get(), ptr, *message);
+          copy_message = MessageUniquePtr(ptr, deleter);
+
+          subscription->provide_intra_process_message(std::move(copy_message));
+        }
+      } else if (is_serialized_publisher) {
+        // publisher provides a serialized message, while subscriber expects a ROS message
+        provide_serialized_intra_process_message(subscription_base, *message);
       } else {
-        // Copy the message since we have additional subscriptions to serve
-        MessageUniquePtr copy_message;
-        Deleter deleter = message.get_deleter();
-        auto ptr = MessageAllocTraits::allocate(*allocator.get(), 1);
-        MessageAllocTraits::construct(*allocator.get(), ptr, *message);
-        copy_message = MessageUniquePtr(ptr, deleter);
-
-        subscription->provide_intra_process_message(std::move(copy_message));
+        // publisher provides a ROS message, while subscriber expects a serialized message
+        rclcpp::Serialization<MessageT> serialization;
+        auto subscription = std::static_pointer_cast<
+          rclcpp::experimental::SubscriptionIntraProcess<SerializedMessage>
+          >(subscription_base);
+
+        auto serialized_message = std::make_unique<SerializedMessage>();
+        serialization.serialize_message(*message, *serialized_message);
+
+        subscription->provide_intra_process_message(std::move(serialized_message));
       }
     }
   }
 
+  template<typename T>
+  static std::enable_if_t<!rclcpp::is_serialized_message_class<T>::value>
+  provide_serialized_intra_process_message(
+    rclcpp::experimental::SubscriptionIntraProcessBase::SharedPtr & subscription,
+    const T & serialized_message)
+  {
+    (void)subscription;
+    (void)serialized_message;
+    // prevent call if actual message type does not match
```

Comment:
could a `static_assert` be added here?

Reply:
I also don't think that these two static methods have to be introduced here. I think the `static_assert` could be part of the function above.

Reply:
I agree with @Karsten1987 on this.